### PR TITLE
support the mixed  image and text watermark feature

### PR DESCRIPTION
**Pull request by tingxin**
support the order of text and image in mixed watermark
support the align of text and image in mixed watermark
support the interval of text and image in mixed watermark

plus:
support x, y position and vertical offset for image watermark

### CODE REVIEW

File: source/new-image-handler/src/processor/image/watermark.ts [hunk 1]
```diff
@@ -14,6 +13,12 @@ export interface WatermarkOpts extends IActionOpts {
   color: string; // 文字颜色
   image: string; // img 水印URL
   auto: boolean; // 自动调整水印图片大小以适应背景
+  x?: number | undefined; // 图文水印的x位置
+  y?: number | undefined; // 图文水印的y位置
```
Comment (wchaws) [inline source/new-image-handler/src/processor/image/watermark.ts:17]:
```ts
x?: number
y?: number
```
即可

File: source/new-image-handler/src/processor/image/watermark.ts [hunk 2]
```diff
@@ -22,11 +27,21 @@ interface WatermarkTextOpts extends IActionOpts {
   height: number;
 }
 
+interface WatermarkImgOpts extends IActionOpts {
+  x?: number | undefined;
+  y?: number | undefined;
```
Comment (wchaws) [inline source/new-image-handler/src/processor/image/watermark.ts:32]: 同上

File: source/new-image-handler/src/processor/image/watermark.ts [hunk 3]
```diff
@@ -169,7 +204,39 @@ export class WatermarkAction implements IImageAction {
         watermarkImg = watermarkImg.resize(width, height);
       }
     }
+    const pos = this.calculateImgPos(opt, metadata, markMetadata);
     const bt = await watermarkImg.toBuffer();
+    let overlay: sharp.OverlayOptions = { input: bt, tile: opt.fill, gravity: opt.g, top: pos.y, left: pos.x };
```
Comment (wchaws) [inline source/new-image-handler/src/processor/image/watermark.ts:209]: `const overlay` ？

File: source/new-image-handler/src/processor/image/watermark.ts [hunk 4]
```diff
@@ -169,7 +204,39 @@ export class WatermarkAction implements IImageAction {
...
+  }
+
+  async mixedWaterMark(ctx: IImageContext, opt: WatermarkOpts): Promise<void> {
+    const bs = ctx.bufferStore;
+    const textOpt = this.calculateTextSize(opt.text, opt.size);
+    const svg = this.textSvgStr(opt, textOpt);
+    const svgBytes = Buffer.from(svg);
+
+    const watermarkImgBuffer = (await bs.get(opt.image)).buffer;
+    let watermarkImg = sharp(watermarkImgBuffer).png();
```
Comment (wchaws) [inline source/new-image-handler/src/processor/image/watermark.ts:220]: const ?

File: source/new-image-handler/src/processor/image/watermark.ts [hunk 5]
```diff
@@ -169,7 +204,39 @@ export class WatermarkAction implements IImageAction {
...
+    const imgMetadata = await watermarkImg.metadata();
+    let imgW = imgMetadata.width ? imgMetadata.width : 0;
+    let imgH = imgMetadata.height ? imgMetadata.height : 0;
```
Comment (wchaws) [inline source/new-image-handler/src/processor/image/watermark.ts:223]: const ?

**Review by wchaws — CHANGES_REQUESTED** [review]

**Comment by tingxin** [comment]
thank  your good suggestion!

**Pull request merged**